sled-agent/src/vmm_reservoir.rs

```diff
@@ -0,0 +1,255 @@
+    /// busy handling another request.
+    //
+    // It's anticipated this will be used to change the reservation
+    #[allow(unused)]
+    pub async fn set_reservoir_size(
+        &self,
+        mode: ReservoirMode,
+    ) -> Result<(), Error> {
+        let (tx, rx) = oneshot::channel();
+        let msg = ReservoirManagerMsg::SetReservoirSize { mode, reply_tx: tx };
+        if let Err(e) = self.tx.try_send(msg) {
+            let (e, is_shutdown) = match e {
+                flume::TrySendError::Full(e) => (e, false),
+                flume::TrySendError::Disconnected(e) => (e, true),
+            };
+            let ReservoirManagerMsg::SetReservoirSize { mode, .. } = e;
+            if is_shutdown {
+                return Err(Error::Shutdown(mode));
+            } else {
+                return Err(Error::Busy(mode));
+            }
+        }
+        rx.await.map_err(|_| Error::ReplySenderDropped)?
+    }
+}
+
+/// Manage the VMM reservoir in a background thread
+pub struct VmmReservoirManager {
+    reservoir_size: Arc<AtomicU64>,
+    rx: flume::Receiver<ReservoirManagerMsg>,
+    log: Logger,
+}
+
+impl VmmReservoirManager {
+    pub fn run(
+        self,
+        hardware_manager: HardwareManager,
+        reservoir_mode: ReservoirMode,
+    ) {
+        match reservoir_mode {
+            ReservoirMode::None => warn!(self.log, "Not using VMM reservoir"),
+            ReservoirMode::Size(0) | ReservoirMode::Percentage(0) => {
+                warn!(
+                    self.log,
+                    "Not using VMM reservoir (size 0 bytes requested)"
+                )
+            }
+            _ => {
+                if let Err(e) =
+                    self.set_reservoir_size(&hardware_manager, reservoir_mode)
+                {
+                    error!(self.log, "Failed to setup VMM reservoir: {e}");
+                }
+            }
+        }
+
+        while let Ok(msg) = self.rx.recv() {
+            let ReservoirManagerMsg::SetReservoirSize { mode, reply_tx } = msg;
+            match self.set_reservoir_size(&hardware_manager, mode) {
+                Ok(()) => {
+                    let _ = reply_tx.send(Ok(()));
+                }
+                Err(e) => {
+                    error!(self.log, "Failed to setup VMM reservoir: {e}");
+                    let _ = reply_tx.send(Err(e));
+                }
+            }
+        }
+    }
+
+    /// Sets the VMM reservoir to the requested percentage of usable physical
+    /// RAM or to a size in MiB. Either mode will round down to the nearest
+    /// aligned size required by the control plane.
+    pub fn set_reservoir_size(
+        &self,
+        hardware: &sled_hardware::HardwareManager,
+        mode: ReservoirMode,
+    ) -> Result<(), Error> {
+        let hardware_physical_ram_bytes = hardware.usable_physical_ram_bytes();
+        let req_bytes = match mode {
+            ReservoirMode::None => return Ok(()),
```

Comment:
This may be a useless nit, but this looks a little confusing to me - if I call `set_reservoir_size` with something other than `None`, that succeeds, then call it with `None`, that will claim to succeed without "turning off" the reservoir (if that concept even makes sense - I guess it'd be "size it back down to 0"?). I'm a little tempted to propose something like:

* Remove the `ReservoirMode::None` variant
* Make `ReservoirMode::from_config` return `Option<ReservoirMode>`
* Make `spawn()` and `run()` take `Option<ReservoirMode>`, but keep both `set_size` functions accepting only `ReservoirMode`

I think this would allow you to start up with no reservoir and later add one, but statically prevents starting up with a reservoir and then later trying to go to `Mode::None`.

Reply:
Great suggestion. Done in [bdea2f6](https://github.com/oxidecomputer/omicron/pull/5124/commits/bdea2f6c01707ad7b9323a9645f1e15e93e96104)